Fixes and interface updates

wwoods opened the pull request:
As discussed, here's a pull request composed of:

* document_classes namespace now imported into the `pubplot` object by default (accessible via e.g. `import pubplot; pubplot.document_classes.ieee_jrnl`).
* IEEE documents correctly refer to the `times` package.
* LaTeX packages specified in `Document`'s arguments are now correctly added to the matplotlib preamble.
* Specifying `font.size` in style will by default override other unspecified font sizes.
* `Figure` now supports `add_gridspec` and `add_subplots`; `Document.figure()` by default does not create any axes, which matches matplotlib's behavior.
* `tight_layout` and `canvas.print_figure` parameters have been fixed to ensure correct font sizes are maintained on save.

On a side note, I see why you have `documentclass` and `document_options`... because of PyLatex.  Too bad (inconsistent underscores), but probably best to follow their suit.

Addresses issue #1

wwoods committed:
Use 'times' package for IEEE; import document_classes

wwoods committed:
Passing font.size overrides other font styles

wwoods committed:
Fixes for latex packages; Figure updates

* Figure now supports GridSpec via add_gridspec() and add_subplot().
* LaTeX packages now appropriately added to preamble.
* Margins now correctly trimmed in a size-preserving manner.

wwoods committed:
Document.temporary_style() for temporary styles

wwoods commented:
Note: I also added `Document.temporary_style()` for temporary styles, which wasn't something supported previously.

wwoods commented:
Missed `w_pad` and `h_pad` on first go-around for figure's `tight_layout`.  Fixed the commit for clean history.

wwoods committed:
Journals usually recommend 300-600 DPI

hsadok commented:
Hi Walt,

Thank you very much! I am *really* glad with your contribution.

> * document_classes namespace now imported into the `pubplot` object by default (accessible via e.g. `import pubplot; pubplot.document_classes.ieee_jrnl`).

This is very useful. I wonder what else should be imported in the `pubplot` object.

> * IEEE documents correctly refer to the `times` package.

Great! I didn't have a problem with the `times` package when I was using this but I see no problem in adding it.

> * LaTeX packages specified in `Document`'s arguments are now correctly added to the matplotlib preamble.

This is essential indeed. Thank you for implementing it.

> * Specifying `font.size` in style will by default override other unspecified font sizes.

I *loved* your solution here, if the user specifies only the `font.size` you override all the other sizes, but you also respect if the user decides to change a specific size.

> * `Figure` now supports `add_gridspec` and `add_subplots`; `Document.figure()` by default does not create any axes, which matches matplotlib's behavior.

I always use figures with axes. So I thought it would be simpler to use the `Document.figure()` method if the return object came with axes "out of the box." But since this may bring limitations, I understand your modifications. Also matching matplotlib's behavior is likely a good call here.

About the other `Figure` methods. I understand that the reason you have implemented `add_subplots` is likely to ensure that it returns a `PubAxes` object. However, I am not entirely sure `add_gridspec` is needed here. Doesn't the fallback to the matplotlib Figure object work for this method? 

> * `tight_layout` and `canvas.print_figure` parameters have been fixed to ensure correct font sizes are maintained on save.

Cool! This avoids some annoying margins that matplotlib adds by default. At the moment I was avoiding this problem by adding a `savefig.pad_inches': 0.0` to the style. But this solution looks more definitive. 

Just one caveat. I ran a quick test with the [simple_plot.py](https://github.com/hugombarreto/pubplot/blob/master/examples/simple_plot.py) and the part with subplots is raising an exception (`numpy.linalg.LinAlgError: Singular matrix`). I suspect it has something to do with these changes but I am not entirely sure. If you have any clue on what this might be please let me know. Otherwise I plan to investigate more deeply during the weekend.

> On a side note, I see why you have `documentclass` and `document_options`... because of PyLatex. Too bad (inconsistent underscores), but probably best to follow their suit.

Yep! That was the reason indeed. 

> Note: I also added `Document.temporary_style()` for temporary styles, which wasn't something supported previously.

This is very useful! Thank you for that as well.

Again, I am really glad with your contributions. I've been using this thing alone for about one year and it is really nice to know that it is now being useful to someone else.

I will merge your request as soon as I figure out the above issues.

wwoods committed:
Document.subfigures() shortcut, fixed tests

wwoods commented:
> Thank you very much! I am _really_ glad with your contribution.

Happy to help!

> I wonder what else should be imported in the `pubplot` object.

When in doubt, don't import anything on the main object.  `document_classes` just gets used basically every time, so I think it makes sense to be imported by default.  I generally prefer a lean module import list though... makes it easier to sort through.  I think at the moment `Document` and `document_classes` are the only things that users of this libraries need to refer to directly.

> About the other `Figure` methods. I understand that the reason you have implemented `add_subplots` is likely to ensure that it returns a `PubAxes` object. However, I am not entirely sure `add_gridspec` is needed here. Doesn't the fallback to the matplotlib Figure object work for this method?

Ah, yes, failed to notice that RCParamWrapper implemented __getattr__.  Fixed.  Also changed `Document.subfigures()` to have a defaults `nrows=1, ncols=1` and implement squeezing so that you can use a shortcut like `fig, ax = d.subfigures()`.

> Just one caveat. I ran a quick test with the [simple_plot.py](https://github.com/hugombarreto/pubplot/blob/master/examples/simple_plot.py) and the part with subplots is raising an exception (`numpy.linalg.LinAlgError: Singular matrix`). I suspect it has something to do with these changes but I am not entirely sure. If you have any clue on what this might be please let me know. Otherwise I plan to investigate more deeply during the weekend.

I see.  I just fixed simple_plot.py for python 3 / the latest changes.  I do not get a "Singular matrix" error; perhaps you need to update your numpy?

> Again, I am really glad with your contributions. I've been using this thing alone for about one year and it is really nice to know that it is now being useful to someone else.

Thanks for the quick response on open-sourcing it!  Much appreciated.

hsadok commented:
Thank you again for your changes.

I have merged your pull request into a `dev` branch.

It seems that it works perfectly with python3. The problem I have noticed only appears when using python2.7. I will see if can figure it out but I am not sure if the burden of supporting python2.7 will be worth in the long run.
